corehq/motech/fhir/matchers.py

```diff
@@ -0,0 +1,245 @@
+        return sum(self.iter_weights(candidate))
+
+    def iter_weights(self, candidate):
+        for pw in self.property_weights:
+            is_match = pw.method.is_match(self.resource, candidate)
+            if is_match is None:
+                # method was unable to compare values
+                continue
+            yield pw.weight if is_match else -1 * pw.negative_weight
+
+
+class ComparisonMethod:
+    """
+    A method of comparing resource properties.
+    """
+
+    def __init__(self, jsonpath: str):
+        self.jsonpath = jsonpath_parse(jsonpath)
+
+    def is_match(self, resource: dict, candidate: dict) -> Optional[bool]:
+        a = simplify_list([x.value for x in self.jsonpath.find(resource)])
+        b = simplify_list([x.value for x in self.jsonpath.find(candidate)])
+        if a is None or b is None:
+            return None
+        return self.compare(a, b)
+
+    @staticmethod
+    def compare(a, b) -> bool:
+        raise NotImplementedError
+
+
+class IsEqual(ComparisonMethod):
+
+    @staticmethod
+    def compare(a, b):
+        return a == b
+
+
+class GivenName(ComparisonMethod):
+
+    @staticmethod
+    def compare(a, b):
+        return any(a_name == b_name for a_name, b_name in zip(a, b))
+
+
+class Age(ComparisonMethod):
+
+    @staticmethod
+    def compare(a, b):
+        max_days = 364
+        return le_days_diff(max_days, a, b)
+
+
+class OrganizationName(ComparisonMethod):
+
+    @staticmethod
+    def compare(a, b):
+        percent = 0.2
+        return le_levenshtein_percent(percent, a.lower(), b.lower())
+
+
+class NegativeIdentifier(ComparisonMethod):
+    """
+    Returns False if the identifier system is the same, and the
+    identifier value is not close.
+
+    Used for giving a negative score to candidates with different IDs
+    from the same system.
+    """
+
+    @staticmethod
+    def compare(a, b):
+        system_a, value_a = a.split('|')
+        system_b, value_b = b.split('|')
+        return not (
+            system_a == system_b
+            and not le_levenshtein_percent(0.2, value_a, value_b)
+        )
+
+
+def _value_ge_zero(instance, attrib, value):
+    return value >= 0
+
+
+@attr.s
+class PropertyWeight:
+    """
+    Associates a matching property with a weight
+    """
+    jsonpath = attr.ib(type=str)
+    weight = attr.ib(type=Decimal, validator=_value_ge_zero)
+    method_class = attr.ib(type=Type[ComparisonMethod], default=IsEqual)
+    # Score negatively if values are different
+    negative_weight = attr.ib(type=Decimal,
+                              default=Decimal('0'),
+                              validator=_value_ge_zero)
+
+    @property
+    def method(self):
+        return self.method_class(self.jsonpath)
+
+
+class PersonMatcher(ResourceMatcher):
+    """
+    Finds matching FHIR Persons
+    """
+    property_weights = [
+        PropertyWeight(
+            '$.identifier[0].system + "|" + $.identifier[0].value',
+            Decimal('0.8'),
```

Comment:
I'm applying a _rough_ heuristic: The weight is inversely proportional to the chance of a false positive; The chances of matching a date of birth are high, so that gets a low weight. The chances of a false positive for a human-entered ID are low, but not 0, because humans. The chances of a false positive on a system-generated ID that is not entered by a human (like a CommCare case ID) is very close to 0, so it gets a high weight.

And by "rough" I mean I'm thumbsucking. So *definitely* room for improvement, ideally by analysing a real/realistic sample.